### PR TITLE
JENKINS-34254 - RequirePOST form not including crumb

### PR DESCRIPTION
**Pull request by kzantow**
Stapler generates a form for methods annotated with `@RequirePOST` but they do not work with CSRF enabled. This alters the form to include a valid crumb, so they function as expected.

This fixes: https://issues.jenkins-ci.org/browse/JENKINS-34254

### CODE REVIEW

**Comment by kzantow** [comment]
@reviewbybees

File: core/src/main/java/hudson/security/csrf/CrumbFilter.java [hunk 1]
```diff
@@ -41,6 +50,39 @@ public CrumbIssuer getCrumbIssuer() {
...
+
+    private void addRequirePostCrumbHandler(WebApp webApp) {
+        webApp.getResponseRenderers().add(0, new HttpResponseRenderer() {
+            @Override
+            public boolean generateResponse(StaplerRequest req, StaplerResponse rsp, Object node, Object response) throws IOException, ServletException {
+                if (response instanceof HttpResponseException) {
+                    Class<?> c = response.getClass();
+                    // The only response with enclosing class of RequirePOST.Processor is the 'try post'
+                    // response, which we need ot modify to be crumb-aware
+                    if (RequirePOST.Processor.class.equals(c.getEnclosingClass())) {
```
Comment (jglick) [inline core/src/main/java/hudson/security/csrf/CrumbFilter.java]: Better to check `c.getName().equals("…")` I think.

File: core/src/main/java/hudson/security/csrf/StaplerResponseWrapper.java [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+import net.sf.json.JsonConfig;
+
+/**
+ * A basic wrapper for a StaplerResponse, e.g. in order to override some method.
+ * This simply delegates to the wrapped instance for everything and is not useful
+ * by itself.
+ * 
+ * FIXME: move to Stapler core
+ * 
+ * @since TODO
```
Comment (jglick) [inline core/src/main/java/hudson/security/csrf/StaplerResponseWrapper.java]: Make this package-private please, so it does not become an API.

**Comment by jglick** [comment]
Works, but it would seem a lot cleaner to have an API in Stapler which allows the application to customize the behavior in this way.

**Comment by kzantow** [comment]
@jglick  Yes, hence the FIXME ; I'll be happy to file a patch against stapler and update the dependencies, if you prefer and we can get it incorporated quickly (this was originally going to be a 2.0 patch, which would have not been able to bump the stapler version, I think)

**Comment by jglick** [comment]
> if […] we can get it incorporated quickly

If an upstream PR passes review, cutting a release is easy enough.

**Comment by kzantow** [comment]
@jglick agreed this is the correct thing to do, here's the stapler PR: https://github.com/stapler/stapler/pull/73

**Comment by daniel-beck** [comment]
Assuming this is work in progress and will be adapted after the change in Stapler makes it in.

**Comment by kzantow** [comment]
@daniel-beck well, it works fine as-is, but yes the better change is in stapler, then adapt this PR

**Comment by andresrc** [comment]
:bee: though if we can get https://github.com/stapler/stapler/pull/73 in better

**Comment by oleg-nenashev** [comment]
@kzantow Any plans to update it?

**Comment by kzantow** [comment]
@oleg-nenashev updated to latest Stapler that contains https://github.com/stapler/stapler/pull/73 , addressed @jglick 's comment checking `c.getName().equals(...`

**Comment by oleg-nenashev** [comment]
Looks good to me

**Comment by kzantow** [comment]
Looks like the stapler upgrade might have introduced a regression, unclear to me; will try to debug the issue in the near future...

**Comment by daniel-beck** [comment]
@kzantow What is the status of this PR?

**Comment by oleg-nenashev** [comment]
@kzantow ping

**Comment by oleg-nenashev** [comment]
any updates?

**Comment by daniel-beck** [comment]
This would be so awesome.

**Comment by daniel-beck** [comment]
@oleg-nenashev @jglick @andresrc PR has been updated. Please re-review.

**Comment by jglick** [comment]
I think it would be better to handle this directly in Stapler with a proper extension point, rather than monkeypatching the HTTP response.

**Comment by daniel-beck** [comment]
@jglick This is already in Stapler for months, so that comment is a bit late, no?

**Comment by jglick** [comment]
> This is already in Stapler for months

No a generic response wrapper is in Stapler. I meant for the actual `POST` handler to allow the `WebApp` to specify a custom form.

**Comment by daniel-beck** [comment]
In any case, we need a solution for this. Otherwise `@RequirePOST` is pointless and we could just use `@POST`.

**Pull request closed without merging**